<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><ul class="toc-item"><li><span><a href="#Общее-впечатление" data-toc-modified-id="Общее-впечатление-0.1"><span class="toc-item-num">0.1&nbsp;&nbsp;</span><font color="orange">Общее впечатление</font></a></span></li><li><span><a href="#Общее-впечатление-(ревью-2)" data-toc-modified-id="Общее-впечатление-(ревью-2)-0.2"><span class="toc-item-num">0.2&nbsp;&nbsp;</span><font color="orange">Общее впечатление (ревью 2)</font></a></span></li></ul></li><li><span><a href="#Постановка-задачи" data-toc-modified-id="Постановка-задачи-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Постановка задачи</a></span><ul class="toc-item"><li><span><a href="#Цель-проекта" data-toc-modified-id="Цель-проекта-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Цель проекта</a></span></li><li><span><a href="#Задачи-проекта" data-toc-modified-id="Задачи-проекта-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>Задачи проекта</a></span></li></ul></li><li><span><a href="#Подготовка-данных" data-toc-modified-id="Подготовка-данных-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Подготовка данных</a></span><ul class="toc-item"><li><span><a href="#Заполнени-пропусков" data-toc-modified-id="Заполнени-пропусков-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Заполнени пропусков</a></span></li><li><span><a href="#Удаление-незначащих-признаков" data-toc-modified-id="Удаление-незначащих-признаков-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Удаление незначащих признаков</a></span></li><li><span><a href="#Прямое-кодирование-категориальных-признаков" data-toc-modified-id="Прямое-кодирование-категориальных-признаков-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Прямое кодирование категориальных признаков</a></span></li><li><span><a href="#Выводы" data-toc-modified-id="Выводы-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>Выводы</a></span></li></ul></li><li><span><a href="#Исследование-задачи" data-toc-modified-id="Исследование-задачи-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Исследование задачи</a></span><ul class="toc-item"><li><span><a href="#Выделение-признаков" data-toc-modified-id="Выделение-признаков-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Выделение признаков</a></span></li><li><span><a href="#Деление-на-выборки" data-toc-modified-id="Деление-на-выборки-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Деление на выборки</a></span></li><li><span><a href="#Масштабирование-признаков" data-toc-modified-id="Масштабирование-признаков-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Масштабирование признаков</a></span></li><li><span><a href="#Дерево-решений" data-toc-modified-id="Дерево-решений-3.4"><span class="toc-item-num">3.4&nbsp;&nbsp;</span>Дерево решений</a></span></li><li><span><a href="#Случайный-лес" data-toc-modified-id="Случайный-лес-3.5"><span class="toc-item-num">3.5&nbsp;&nbsp;</span>Случайный лес</a></span></li><li><span><a href="#Логистическая-регрессия" data-toc-modified-id="Логистическая-регрессия-3.6"><span class="toc-item-num">3.6&nbsp;&nbsp;</span>Логистическая регрессия</a></span></li><li><span><a href="#Выводы." data-toc-modified-id="Выводы.-3.7"><span class="toc-item-num">3.7&nbsp;&nbsp;</span>Выводы.</a></span></li></ul></li><li><span><a href="#Борьба-с-дисбалансом" data-toc-modified-id="Борьба-с-дисбалансом-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Борьба с дисбалансом</a></span><ul class="toc-item"><li><span><a href="#Учёт-весов." data-toc-modified-id="Учёт-весов.-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Учёт весов.</a></span></li><li><span><a href="#Увеличение-выборки" data-toc-modified-id="Увеличение-выборки-4.2"><span class="toc-item-num">4.2&nbsp;&nbsp;</span>Увеличение выборки</a></span></li><li><span><a href="#Уменьшение-выборки" data-toc-modified-id="Уменьшение-выборки-4.3"><span class="toc-item-num">4.3&nbsp;&nbsp;</span>Уменьшение выборки</a></span></li><li><span><a href="#Выводы" data-toc-modified-id="Выводы-4.4"><span class="toc-item-num">4.4&nbsp;&nbsp;</span>Выводы</a></span></li></ul></li><li><span><a href="#Тестирование-модели" data-toc-modified-id="Тестирование-модели-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Тестирование модели</a></span></li><li><span><a href="#Общие-выводы" data-toc-modified-id="Общие-выводы-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Общие выводы</a></span></li><li><span><a href="#Чек-лист-готовности-проекта" data-toc-modified-id="Чек-лист-готовности-проекта-7"><span class="toc-item-num">7&nbsp;&nbsp;</span>Чек-лист готовности проекта</a></span></li></ul></div>

# Отток клиентов
## Постановка задачи

Из «Бета-Банка» стали уходить клиенты. Каждый месяц. Немного, но заметно. Банковские маркетологи посчитали: сохранять текущих клиентов дешевле, чем привлекать новых.

Нужно спрогнозировать, уйдёт клиент из банка в ближайшее время или нет. Вам предоставлены исторические данные о поведении клиентов и расторжении договоров с банком. 

Постройте модель с предельно большим значением *F1*-меры. Чтобы сдать проект успешно, нужно довести метрику до 0.59. Проверьте *F1*-меру на тестовой выборке самостоятельно.

Дополнительно измеряйте *AUC-ROC*, сравнивайте её значение с *F1*-мерой.

Источник данных: [https://www.kaggle.com/barelydedicated/bank-customer-churn-modeling](https://www.kaggle.com/barelydedicated/bank-customer-churn-modeling)

### Цель проекта

Спрогнозировать, уйдёт клиент из банка в ближайшее время или нет

### Задачи проекта
 * Исследовать баланс классов и обучить модель без учёта дисбаланса
 * Улучшите качество модели, учитывая дисбаланс классов.
 * Обучите разные модели и найдите лучшую
 * Построить модель с предельно большим значением F1-меры (не менне 0.59)
 * Измерить AUC-ROC и сравнивать её значение с F1-мерой.

 

## Подготовка данных

Импортируем библиотеки

In [1]:
import pandas as pd
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
from sklearn import tree
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import f1_score, roc_auc_score
from sklearn.utils import shuffle

Читаем файл

In [2]:
df=pd.read_csv('/datasets/Churn.csv')

Прочитался без ошибок. Вывеедм информацию о таблице

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 14 columns):
RowNumber          10000 non-null int64
CustomerId         10000 non-null int64
Surname            10000 non-null object
CreditScore        10000 non-null int64
Geography          10000 non-null object
Gender             10000 non-null object
Age                10000 non-null int64
Tenure             9091 non-null float64
Balance            10000 non-null float64
NumOfProducts      10000 non-null int64
HasCrCard          10000 non-null int64
IsActiveMember     10000 non-null int64
EstimatedSalary    10000 non-null float64
Exited             10000 non-null int64
dtypes: float64(3), int64(8), object(3)
memory usage: 1.1+ MB


Пропуски есть только в столбце `Tenure`. 

Согласно заданию в столбах следующие признаки:
* `RowNumber` — индекс строки в данных
* `CustomerId` — уникальный идентификатор клиента
* `Surname` — фамилия
* `CreditScore` — кредитный рейтинг
* `Geography` — страна проживания
* `Gender` — пол
* `Age` — возраст
* `Tenure` — сколько лет человек является клиентом банка
* `Balance` — баланс на счёте
* `NumOfProducts` — количество продуктов банка, используемых клиентом
* `HasCrCard` — наличие кредитной карты
* `IsActiveMember` — активность клиента
* `EstimatedSalary` — предполагаемая зарплата

Целевой признак
* `Exited` — факт ухода клиента

Выведем первые 5 строк

In [4]:
df.head()

RowNumber  CustomerId   Surname  CreditScore Geography  Gender  Age  \
0          1    15634602  Hargrave          619    France  Female   42   
1          2    15647311      Hill          608     Spain  Female   41   
2          3    15619304      Onio          502    France  Female   42   
3          4    15701354      Boni          699    France  Female   39   
4          5    15737888  Mitchell          850     Spain  Female   43   

   Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0     2.0       0.00              1          1               1   
1     1.0   83807.86              1          0               1   
2     8.0  159660.80              3          1               0   
3     1.0       0.00              2          0               0   
4     2.0  125510.82              1          1               1   

   EstimatedSalary  Exited  
0        101348.88       1  
1        112542.58       0  
2        113931.57       1  
3         93826.63       0  
4         79084.10       0

И выведем к-во уникальных значений

In [5]:
df.nunique()

RowNumber          10000
CustomerId         10000
Surname             2932
CreditScore          460
Geography              3
Gender                 2
Age                   70
Tenure                11
Balance             6382
NumOfProducts          4
HasCrCard              2
IsActiveMember         2
EstimatedSalary     9999
Exited                 2
dtype: int64

### Заполнени пропусков

Так как в исследовании планируем использовать дерево решений, случайный лес и логистическую регрессию, которые не умеют работать с пропусками, то необходимо заполнить пропуски пропусками. Заполним пропуски 0 в столбце `Tenure` и добавим дополнительный столбец `TenureNA` в котором будем писать True, если был попуск в столбце `Tenure`.

In [6]:
df['TenureNA'] = df['Tenure'].isna()

In [7]:
df['Tenure'] = df['Tenure'].fillna(0)

In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 15 columns):
RowNumber          10000 non-null int64
CustomerId         10000 non-null int64
Surname            10000 non-null object
CreditScore        10000 non-null int64
Geography          10000 non-null object
Gender             10000 non-null object
Age                10000 non-null int64
Tenure             10000 non-null float64
Balance            10000 non-null float64
NumOfProducts      10000 non-null int64
HasCrCard          10000 non-null int64
IsActiveMember     10000 non-null int64
EstimatedSalary    10000 non-null float64
Exited             10000 non-null int64
TenureNA           10000 non-null bool
dtypes: bool(1), float64(3), int64(8), object(3)
memory usage: 1.1+ MB


Теперь пропусков нет

### Удаление незначащих признаков

Видим, что столбец `RowNumber` фактически дублирует индекс. Этот столбец можно дропнуть. Также не имеющие значимости признаки - это `СustomerId` и	`Surname`. Маловероятно, что фамилия клиента влияла на то, уйдёт он или нет.

In [9]:
df = df.drop(['RowNumber','CustomerId', 'Surname'], axis = 1 )

In [10]:
df.head()

CreditScore Geography  Gender  Age  Tenure    Balance  NumOfProducts  \
0          619    France  Female   42     2.0       0.00              1   
1          608     Spain  Female   41     1.0   83807.86              1   
2          502    France  Female   42     8.0  159660.80              3   
3          699    France  Female   39     1.0       0.00              2   
4          850     Spain  Female   43     2.0  125510.82              1   

   HasCrCard  IsActiveMember  EstimatedSalary  Exited  TenureNA  
0          1               1        101348.88       1     False  
1          0               1        112542.58       0     False  
2          1               0        113931.57       1     False  
3          0               0         93826.63       0     False  
4          1               1         79084.10       0     False

### Прямое кодирование категориальных признаков

Некоторые столбцы не имеют числового типа. Это `Geography`, `Gender`. Ранее показали, что уникальных значений у `Geography` 3, а у `Gender` - 2. Имеет смысл их заменить дамми-переменными. Для уменьшения к-ва столбцов в дамми-переменных первый столбец дропнем.

In [11]:
df = pd.get_dummies(df, drop_first=True)

In [12]:
df.head()

CreditScore  Age  Tenure    Balance  NumOfProducts  HasCrCard  \
0          619   42     2.0       0.00              1          1   
1          608   41     1.0   83807.86              1          0   
2          502   42     8.0  159660.80              3          1   
3          699   39     1.0       0.00              2          0   
4          850   43     2.0  125510.82              1          1   

   IsActiveMember  EstimatedSalary  Exited  TenureNA  Geography_Germany  \
0               1        101348.88       1     False                  0   
1               1        112542.58       0     False                  0   
2               0        113931.57       1     False                  0   
3               0         93826.63       0     False                  0   
4               1         79084.10       0     False                  0   

   Geography_Spain  Gender_Male  
0                0            0  
1                1            0  
2                0            0  
3                0            0  
4                1            0

### Выводы

Файл открылся без ошибок. Столбцы соответствуют заявленным. Пропуски были в одном столбце. Их земенили и ввели доп.признак. Категориальные признаки закодировали по OHE.

## Исследование задачи

### Выделение признаков
Для начала выделим признаки, на основании которых будет производиться кобучение. Это все признаки, кроме `Exited`.

In [13]:
features = df.drop('Exited', axis=1) 

Целевой показатель - `Exited`, который определяет ушёл клиент или нет

In [14]:
target = df['Exited']

Посмотрим на баланс классов

In [15]:
target.value_counts(normalize=True)

0    0.7963
1    0.2037
Name: Exited, dtype: float64

Видим, что соотношение клиентов которые остались к тем, кто покинул банк примерно 80/20.

### Деление на выборки
Разделим данные в отношении 3:1:1 на тренировочную, валидационную и тестовую выборки. На тестовой и валидационной будем выбирать лучшую модель, а зетем её проверим на тестовой. Отделим сначала тестовую выборку.

In [16]:
features_train, features_test, target_train, target_test= train_test_split(
    features, target, test_size=0.2, random_state = 12345)

Затем из оставшейся выборки `features_train` и `target_train` ещё раз выделим валидационную выборку. Её размер примем как размер тестовой, но с учётом уменьшения размера исходной выборки, параметр `test_size` будет равен 0.2/0.8 = 0.25

In [17]:
features_train, features_valid, target_train, target_valid = train_test_split(
    features_train, target_train, test_size=0.25, random_state = 12345)

Посмотрим, как разделились данные

In [18]:
len(features_train), len(features_valid), len(features_test)

(6000, 2000, 2000)

In [19]:
len(target_train), len(target_valid), len(target_test)

(6000, 2000, 2000)

### Масштабирование признаков

Для корректного обучения модели есть смысл масштабировать небинарные признаки: 	`CreditScore`,	`Age`,	`Tenure`, `Balance`, `EstimatedSalary`. Масштабируем их на тренировочной выборке

In [21]:
features_to_scale = ['CreditScore','Age', 'Tenure', 'Balance', 'EstimatedSalary']
scaler = StandardScaler()
scaler.fit(features_train[features_to_scale])

StandardScaler(copy=True, with_mean=True, with_std=True)

Теперь применим скалер для всех выборок

In [22]:
features_train[features_to_scale] = scaler.transform(features_train[features_to_scale])

In [23]:
features_valid[features_to_scale] = scaler.transform(features_valid[features_to_scale])

In [24]:
features_test[features_to_scale] = scaler.transform(features_test[features_to_scale])

/opt/conda/lib/python3.7/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.
/opt/conda/lib/python3.7/site-packages/pandas/core/indexing.py:494: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self.obj[item] = s


Посмотрим на признаки

In [27]:
features_valid.head()

CreditScore       Age    Tenure   Balance  NumOfProducts  HasCrCard  \
2358     0.175393  0.399118 -1.131841  1.385698              1          0   
8463    -1.299609  0.971741 -0.812608 -1.232442              1          1   
163      0.711757 -0.268942 -0.812608 -1.232442              2          1   
3074    -0.391916  0.494555  0.464325  0.672529              1          1   
5989     0.165078  1.353490  1.741257  0.536522              1          0   

      IsActiveMember  EstimatedSalary  TenureNA  Geography_Germany  \
2358               1        -1.466761     False                  0   
8463               0         0.254415     False                  0   
163                1         0.122863     False                  0   
3074               0         0.585847     False                  1   
5989               0         1.462457     False                  0   

      Geography_Spain  Gender_Male  
2358                0            1  
8463                1            1  
163                 1            0  
3074                0            0  
5989                0            0

### Дерево решений

Воспользуемся `DecisionTreeClassifier` и будем менять глубину для получения наилучшего результата. Циклично выведем F1 и AUC_ROC. Так как для всех моделей планируем определять метрики F1 и AUC_ROC, то для упрощения определим функцию, которая на вход будет принимать модель и признаки, в теле функции будет определять метрики и будет возвращать кортеж f1, auc_roc

In [28]:
def f1_auc(model, target_valid, predictions_valid):
    result = f1_score(target_valid, predictions_valid)
    
    probabilities_valid = model.predict_proba(features_valid)
    probabilities_one_valid = probabilities_valid[:, 1]
    auc_roc = roc_auc_score(target_valid, probabilities_one_valid)
    
    return (result, auc_roc)

Обучим дерево

In [29]:
best_result=0
best_tree_model = None
for i in range(1,30):
    model = tree.DecisionTreeClassifier(max_depth = i, random_state = 12345)
    model.fit(features_train, target_train)
    predictions_valid = model.predict(features_valid)
    res=f1_auc(model, target_valid, predictions_valid)   
    print('depth={:.0f} f1={:>.4f} auc_roc={:>.4f}'.format(i, *res))
   
    if res[0] > best_result:
        best_result = res[0]
        best_tree_model = model

/opt/conda/lib/python3.7/site-packages/sklearn/metrics/classification.py:1437: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no predicted samples.
  'precision', 'predicted', average, warn_for)


depth=1 f1=0.0000 auc_roc=0.6905
depth=2 f1=0.5037 auc_roc=0.7396
depth=3 f1=0.3938 auc_roc=0.7939
depth=4 f1=0.4318 auc_roc=0.8069
depth=5 f1=0.5483 auc_roc=0.8207
depth=6 f1=0.5070 auc_roc=0.8235
depth=7 f1=0.5498 auc_roc=0.8281
depth=8 f1=0.5413 auc_roc=0.8006
depth=9 f1=0.5365 auc_roc=0.7751
depth=10 f1=0.5335 auc_roc=0.7512
depth=11 f1=0.5185 auc_roc=0.7460
depth=12 f1=0.5064 auc_roc=0.7143
depth=13 f1=0.4912 auc_roc=0.6868
depth=14 f1=0.4973 auc_roc=0.6891
depth=15 f1=0.4987 auc_roc=0.6859
depth=16 f1=0.4909 auc_roc=0.6929
depth=17 f1=0.4865 auc_roc=0.6832
depth=18 f1=0.4745 auc_roc=0.6738
depth=19 f1=0.4773 auc_roc=0.6797
depth=20 f1=0.4730 auc_roc=0.6755
depth=21 f1=0.4829 auc_roc=0.6812
depth=22 f1=0.4776 auc_roc=0.6778
depth=23 f1=0.4832 auc_roc=0.6811
depth=24 f1=0.4752 auc_roc=0.6742
depth=25 f1=0.4752 auc_roc=0.6742
depth=26 f1=0.4752 auc_roc=0.6742
depth=27 f1=0.4752 auc_roc=0.6742
depth=28 f1=0.4752 auc_roc=0.6742
depth=29 f1=0.4752 auc_roc=0.6742


Видно, чтo лучшая метрика для дерева с глубиной 7. Но при этом F1 =0.5498, что недотягивает до заданного значения 0.59. После глубины 25 метрика практически не меняется.

In [30]:
best_tree_model.get_params()

{'class_weight': None,
 'criterion': 'gini',
 'max_depth': 7,
 'max_features': None,
 'max_leaf_nodes': None,
 'min_impurity_decrease': 0.0,
 'min_impurity_split': None,
 'min_samples_leaf': 1,
 'min_samples_split': 2,
 'min_weight_fraction_leaf': 0.0,
 'presort': False,
 'random_state': 12345,
 'splitter': 'best'}

### Случайный лес

Построим модель на основании `RandomForestClassifier`. Для этого будем обучать модель на тренировочной выборке и изменять количество деревьев от 0 до 50. Сравним F1 на валидационной выборке и выберем лучший вариант.

In [31]:
best_result=0
best_forest_model = None
for i in range(1,50):
    model = RandomForestClassifier(random_state = 12345, n_estimators = i)
    model.fit(features_train, target_train)
    predictions_valid = model.predict(features_valid)
    res=f1_auc(model, target_valid, predictions_valid)  
    print('n_estimators={:.0f} f1={:>.4f} auc_roc={:>.4f}'.format(i, *res))
    
    if res[0] > best_result:
        best_result = res[0]
        best_forest_model = model

n_estimators=1 f1=0.4477 auc_roc=0.6582
n_estimators=2 f1=0.3792 auc_roc=0.7151
n_estimators=3 f1=0.5000 auc_roc=0.7409
n_estimators=4 f1=0.4860 auc_roc=0.7616
n_estimators=5 f1=0.5175 auc_roc=0.7742
n_estimators=6 f1=0.5122 auc_roc=0.7785
n_estimators=7 f1=0.5303 auc_roc=0.7819
n_estimators=8 f1=0.5182 auc_roc=0.7887
n_estimators=9 f1=0.5469 auc_roc=0.7936
n_estimators=10 f1=0.5277 auc_roc=0.7980
n_estimators=11 f1=0.5581 auc_roc=0.8018
n_estimators=12 f1=0.5413 auc_roc=0.8038
n_estimators=13 f1=0.5617 auc_roc=0.8065
n_estimators=14 f1=0.5340 auc_roc=0.8108
n_estimators=15 f1=0.5514 auc_roc=0.8113
n_estimators=16 f1=0.5372 auc_roc=0.8142
n_estimators=17 f1=0.5432 auc_roc=0.8179
n_estimators=18 f1=0.5367 auc_roc=0.8230
n_estimators=19 f1=0.5437 auc_roc=0.8257
n_estimators=20 f1=0.5340 auc_roc=0.8271
n_estimators=21 f1=0.5443 auc_roc=0.8274
n_estimators=22 f1=0.5366 auc_roc=0.8281
n_estimators=23 f1=0.5420 auc_roc=0.8299
n_estimators=24 f1=0.5390 auc_roc=0.8288
n_estimators=25 f1=0.5501

In [32]:
best_forest_model.get_params()

{'bootstrap': True,
 'class_weight': None,
 'criterion': 'gini',
 'max_depth': None,
 'max_features': 'auto',
 'max_leaf_nodes': None,
 'min_impurity_decrease': 0.0,
 'min_impurity_split': None,
 'min_samples_leaf': 1,
 'min_samples_split': 2,
 'min_weight_fraction_leaf': 0.0,
 'n_estimators': 13,
 'n_jobs': None,
 'oob_score': False,
 'random_state': 12345,
 'verbose': 0,
 'warm_start': False}

Видно, чтo лучшая метрика для леса с кол-вом деревьев 13. Но при этом F1 =0.5617, что тоже недотягивает до заданного значения 0.59.

### Логистическая регрессия

In [33]:
logistic_model = LogisticRegression(random_state = 12345, solver='liblinear')
logistic_model.fit(features_train, target_train)
predictions_valid = logistic_model.predict(features_valid)

print('f1={:>.4f} auc_roc={:>.4f}'.format(*f1_auc(logistic_model, target_valid, predictions_valid)))

f1=0.3008 auc_roc=0.7702


Здесь метрика значительно ниже требуемой

### Выводы. 
Ни одна из моделей не дала желаемого результата. Наиболее близок был случайный лес.

## Борьба с дисбалансом

### Учёт весов.

Первым этапом обучим модели с учётом весов. Укажем, `class_weight`= *balanced* и повторим обучения

In [34]:
best_result=0
best_bal_tree_model = None
for i in range(1,40):
    model = tree.DecisionTreeClassifier(max_depth = i, random_state = 12345, class_weight='balanced')
    model.fit(features_train, target_train)
    predictions_valid = model.predict(features_valid)
    res=f1_auc(model, target_valid, predictions_valid)   
    print('depth={:.0f} f1={:>.4f} auc_roc={:>.4f}'.format(i, *res))
   
    if res[0] > best_result:
        best_result = res[0]
        best_bal_tree_model = model

depth=1 f1=0.4751 auc_roc=0.6899
depth=2 f1=0.4976 auc_roc=0.7382
depth=3 f1=0.4976 auc_roc=0.7909
depth=4 f1=0.5170 auc_roc=0.8065
depth=5 f1=0.5489 auc_roc=0.8188
depth=6 f1=0.5587 auc_roc=0.8089
depth=7 f1=0.5532 auc_roc=0.8005
depth=8 f1=0.5433 auc_roc=0.7850
depth=9 f1=0.5131 auc_roc=0.7529
depth=10 f1=0.4895 auc_roc=0.7247
depth=11 f1=0.4950 auc_roc=0.7250
depth=12 f1=0.4756 auc_roc=0.6972
depth=13 f1=0.4867 auc_roc=0.6988
depth=14 f1=0.4750 auc_roc=0.6850
depth=15 f1=0.4610 auc_roc=0.6765
depth=16 f1=0.4600 auc_roc=0.6710
depth=17 f1=0.4643 auc_roc=0.6702
depth=18 f1=0.4759 auc_roc=0.6746
depth=19 f1=0.4623 auc_roc=0.6654
depth=20 f1=0.4674 auc_roc=0.6684
depth=21 f1=0.4646 auc_roc=0.6665
depth=22 f1=0.4611 auc_roc=0.6639
depth=23 f1=0.4632 auc_roc=0.6649
depth=24 f1=0.4605 auc_roc=0.6635
depth=25 f1=0.4605 auc_roc=0.6635
depth=26 f1=0.4605 auc_roc=0.6635
depth=27 f1=0.4605 auc_roc=0.6635
depth=28 f1=0.4605 auc_roc=0.6635
depth=29 f1=0.4605 auc_roc=0.6635
depth=30 f1=0.4605 auc_

In [35]:
best_bal_tree_model.get_params()

{'class_weight': 'balanced',
 'criterion': 'gini',
 'max_depth': 6,
 'max_features': None,
 'max_leaf_nodes': None,
 'min_impurity_decrease': 0.0,
 'min_impurity_split': None,
 'min_samples_leaf': 1,
 'min_samples_split': 2,
 'min_weight_fraction_leaf': 0.0,
 'presort': False,
 'random_state': 12345,
 'splitter': 'best'}

Метрика лучше - 0,5587. Посмотрим на лес.

In [36]:
best_result=0
best_bal_forest_model = None
for i in range(1,50):
    model = RandomForestClassifier(random_state = 12345, n_estimators = i, class_weight='balanced')
    model.fit(features_train, target_train)
    predictions_valid = model.predict(features_valid)
    
    res =  f1_auc(model, target_valid, predictions_valid)
    print('n_estimators={:.0f} f1={:>.4f} auc_roc={:>.4f}'.format(i, *res))
    
    if res[0] > best_result:
        best_result = res[0]
        best_bal_forest_model = model

n_estimators=1 f1=0.4635 auc_roc=0.6658
n_estimators=2 f1=0.3922 auc_roc=0.7213
n_estimators=3 f1=0.5234 auc_roc=0.7416
n_estimators=4 f1=0.4494 auc_roc=0.7599
n_estimators=5 f1=0.5107 auc_roc=0.7702
n_estimators=6 f1=0.4950 auc_roc=0.7810
n_estimators=7 f1=0.5239 auc_roc=0.7941
n_estimators=8 f1=0.5058 auc_roc=0.8013
n_estimators=9 f1=0.5337 auc_roc=0.8059
n_estimators=10 f1=0.5156 auc_roc=0.8123
n_estimators=11 f1=0.5688 auc_roc=0.8178
n_estimators=12 f1=0.5348 auc_roc=0.8221
n_estimators=13 f1=0.5598 auc_roc=0.8277
n_estimators=14 f1=0.5475 auc_roc=0.8306
n_estimators=15 f1=0.5540 auc_roc=0.8284
n_estimators=16 f1=0.5365 auc_roc=0.8285
n_estimators=17 f1=0.5593 auc_roc=0.8301
n_estimators=18 f1=0.5403 auc_roc=0.8334
n_estimators=19 f1=0.5544 auc_roc=0.8343
n_estimators=20 f1=0.5478 auc_roc=0.8348
n_estimators=21 f1=0.5723 auc_roc=0.8321
n_estimators=22 f1=0.5463 auc_roc=0.8300
n_estimators=23 f1=0.5540 auc_roc=0.8320
n_estimators=24 f1=0.5422 auc_roc=0.8346
n_estimators=25 f1=0.5498

In [37]:
best_bal_forest_model.get_params()

{'bootstrap': True,
 'class_weight': 'balanced',
 'criterion': 'gini',
 'max_depth': None,
 'max_features': 'auto',
 'max_leaf_nodes': None,
 'min_impurity_decrease': 0.0,
 'min_impurity_split': None,
 'min_samples_leaf': 1,
 'min_samples_split': 2,
 'min_weight_fraction_leaf': 0.0,
 'n_estimators': 21,
 'n_jobs': None,
 'oob_score': False,
 'random_state': 12345,
 'verbose': 0,
 'warm_start': False}

Метрика улучшилась - 0,5723. Посмотрим на регрессию

In [38]:
bal_logistic_model = LogisticRegression(random_state = 12345, solver='liblinear', class_weight='balanced')
bal_logistic_model.fit(features_train, target_train)
predictions_valid = bal_logistic_model.predict(features_valid)

print('f1={:>.4f} auc_roc={:>.4f}'.format(*f1_auc(bal_logistic_model, target_valid, predictions_valid)))

f1=0.4755 auc_roc=0.7724


Метрика возрасла, но недостаточно.

### Увеличение выборки

Следующим этапом попробуем использовать апсемплинг. Определим функцию для этого

In [39]:
def upsample(features, target, repeat):
    features_zeros = features[target == 0]
    features_ones = features[target == 1]
    target_zeros = target[target == 0]
    target_ones = target[target == 1]

    features_upsampled = pd.concat([features_zeros] + [features_ones] * repeat)
    target_upsampled = pd.concat([target_zeros] + [target_ones] * repeat)
    features_upsampled, target_upsampled = shuffle(
    features_upsampled, target_upsampled, random_state=12345)
    
    return features_upsampled, target_upsampled
    

Применим функцию с кратностью 4 для увеличения выборки

In [40]:
features_upsampled, target_upsampled = upsample(features_train, target_train, 4)

Посмотрим на баланс классов

In [41]:
target_upsampled.value_counts(normalize=True)

1    0.504919
0    0.495081
Name: Exited, dtype: float64

Количество более-менее одинаково. Обучим дерево

In [42]:
best_result=0
best_up_tree_model = None
for i in range(1,40):
    model = tree.DecisionTreeClassifier(max_depth = i, random_state = 12345)
    model.fit(features_upsampled, target_upsampled)
    predictions_valid = model.predict(features_valid)
    res=f1_auc(model, target_valid, predictions_valid)   
    print('depth={:.0f} f1={:>.4f} auc_roc={:>.4f}'.format(i, *res))
   
    if res[0] > best_result:
        best_result = res[0]
        best_up_tree_model = model

depth=1 f1=0.4751 auc_roc=0.6899
depth=2 f1=0.4976 auc_roc=0.7382
depth=3 f1=0.4976 auc_roc=0.7909
depth=4 f1=0.5170 auc_roc=0.8065
depth=5 f1=0.5489 auc_roc=0.8188
depth=6 f1=0.5587 auc_roc=0.8089
depth=7 f1=0.5532 auc_roc=0.8005
depth=8 f1=0.5424 auc_roc=0.7828
depth=9 f1=0.5081 auc_roc=0.7459
depth=10 f1=0.4851 auc_roc=0.7212
depth=11 f1=0.4829 auc_roc=0.7167
depth=12 f1=0.4752 auc_roc=0.6965
depth=13 f1=0.4803 auc_roc=0.6934
depth=14 f1=0.4610 auc_roc=0.6755
depth=15 f1=0.4729 auc_roc=0.6804
depth=16 f1=0.4635 auc_roc=0.6719
depth=17 f1=0.4697 auc_roc=0.6726
depth=18 f1=0.4733 auc_roc=0.6732
depth=19 f1=0.4544 auc_roc=0.6604
depth=20 f1=0.4535 auc_roc=0.6595
depth=21 f1=0.4622 auc_roc=0.6644
depth=22 f1=0.4638 auc_roc=0.6655
depth=23 f1=0.4609 auc_roc=0.6635
depth=24 f1=0.4636 auc_roc=0.6651
depth=25 f1=0.4636 auc_roc=0.6651
depth=26 f1=0.4636 auc_roc=0.6651
depth=27 f1=0.4636 auc_roc=0.6651
depth=28 f1=0.4636 auc_roc=0.6651
depth=29 f1=0.4636 auc_roc=0.6651
depth=30 f1=0.4636 auc_

Метрика 0.5587, однако всё равно мала. Посмотрим на лес.

In [43]:
best_result=0
best_up_forest_model = None
for i in range(1,50):
    model = RandomForestClassifier(random_state = 123, n_estimators = i)
    model.fit(features_upsampled, target_upsampled)
    predictions_valid = model.predict(features_valid)
    
    res =  f1_auc(model, target_valid, predictions_valid)
    print('n_estimators={:.0f} f1={:>.4f} auc_roc={:>.4f}'.format(i, *res))
    
    if res[0] > best_result:
        best_result = res[0]
        best_up_forest_model = model

n_estimators=1 f1=0.4716 auc_roc=0.6791
n_estimators=2 f1=0.4504 auc_roc=0.7376
n_estimators=3 f1=0.5267 auc_roc=0.7724
n_estimators=4 f1=0.5179 auc_roc=0.7773
n_estimators=5 f1=0.5426 auc_roc=0.7909
n_estimators=6 f1=0.5237 auc_roc=0.7943
n_estimators=7 f1=0.5321 auc_roc=0.7988
n_estimators=8 f1=0.5308 auc_roc=0.8008
n_estimators=9 f1=0.5432 auc_roc=0.8070
n_estimators=10 f1=0.5539 auc_roc=0.8080
n_estimators=11 f1=0.5472 auc_roc=0.8063
n_estimators=12 f1=0.5578 auc_roc=0.8096
n_estimators=13 f1=0.5509 auc_roc=0.8110
n_estimators=14 f1=0.5514 auc_roc=0.8101
n_estimators=15 f1=0.5557 auc_roc=0.8179
n_estimators=16 f1=0.5601 auc_roc=0.8184
n_estimators=17 f1=0.5645 auc_roc=0.8189
n_estimators=18 f1=0.5593 auc_roc=0.8211
n_estimators=19 f1=0.5683 auc_roc=0.8258
n_estimators=20 f1=0.5657 auc_roc=0.8278
n_estimators=21 f1=0.5620 auc_roc=0.8282
n_estimators=22 f1=0.5560 auc_roc=0.8285
n_estimators=23 f1=0.5640 auc_roc=0.8303
n_estimators=24 f1=0.5698 auc_roc=0.8316
n_estimators=25 f1=0.5762

Искомая метрика получена 0,5961. Лучшая модель с к-вом деревьев 40.

In [44]:
best_up_forest_model.get_params()

{'bootstrap': True,
 'class_weight': None,
 'criterion': 'gini',
 'max_depth': None,
 'max_features': 'auto',
 'max_leaf_nodes': None,
 'min_impurity_decrease': 0.0,
 'min_impurity_split': None,
 'min_samples_leaf': 1,
 'min_samples_split': 2,
 'min_weight_fraction_leaf': 0.0,
 'n_estimators': 40,
 'n_jobs': None,
 'oob_score': False,
 'random_state': 123,
 'verbose': 0,
 'warm_start': False}

Посмотрим на регрессию

In [45]:
up_logistic_model = LogisticRegression(random_state = 12345, solver='liblinear')
up_logistic_model.fit(features_upsampled, target_upsampled)
predictions_valid = up_logistic_model.predict(features_valid)

print('f1={:>.4f} auc_roc={:>.4f}'.format(*f1_auc(up_logistic_model, target_valid, predictions_valid)))

f1=0.4741 auc_roc=0.7724


Метрика почти не изменилась.

### Уменьшение выборки

Ещё попробуем уменьшение выборки как средство борьбы с дисбалансом. Объявим функцию

In [46]:
def downsample(features, target, fraction):
    features_zeros = features[target == 0]
    features_ones = features[target == 1]
    target_zeros = target[target == 0]
    target_ones = target[target == 1]
    
    features_sample = features_zeros.sample(frac=fraction, random_state=12345)
    target_sample = target_zeros.sample(frac=fraction, random_state=12345)
    
    features_downsampled = pd.concat([features_sample] + [features_ones])
    target_downsampled = pd.concat([target_sample] + [target_ones])
    
    features_downsampled, target_downsampled = shuffle(
    features_downsampled, target_downsampled, random_state=12345)
    return features_downsampled, target_downsampled

Создадим уменьшенную выборку

In [47]:
features_downsampled, target_downsampled = downsample(features_train, target_train, 0.25)

Посмотрим баланс

In [48]:
target_downsampled.value_counts(normalize=True)

1    0.504971
0    0.495029
Name: Exited, dtype: float64

Обучим дерево

In [49]:
best_result=0
best_down_tree_model = None
for i in range(1,30):
    model = tree.DecisionTreeClassifier(max_depth = i, random_state = 12345)
    model.fit(features_downsampled, target_downsampled)
    predictions_valid = model.predict(features_valid)
    res=f1_auc(model, target_valid, predictions_valid)   
    print('depth={:.0f} f1={:>.4f} auc_roc={:>.4f}'.format(i, *res))
   
    if res[0] > best_result:
        best_result = res[0]
        best_down_tree_model = model

depth=1 f1=0.4751 auc_roc=0.6899
depth=2 f1=0.4976 auc_roc=0.7382
depth=3 f1=0.5118 auc_roc=0.7845
depth=4 f1=0.4957 auc_roc=0.7928
depth=5 f1=0.5328 auc_roc=0.8112
depth=6 f1=0.5390 auc_roc=0.8087
depth=7 f1=0.5441 auc_roc=0.8120
depth=8 f1=0.5170 auc_roc=0.7734
depth=9 f1=0.5009 auc_roc=0.7446
depth=10 f1=0.4948 auc_roc=0.7194
depth=11 f1=0.4723 auc_roc=0.7110
depth=12 f1=0.4930 auc_roc=0.7005
depth=13 f1=0.4908 auc_roc=0.6998
depth=14 f1=0.4719 auc_roc=0.6967
depth=15 f1=0.4745 auc_roc=0.6949
depth=16 f1=0.4704 auc_roc=0.6949
depth=17 f1=0.4541 auc_roc=0.6791
depth=18 f1=0.4658 auc_roc=0.6894
depth=19 f1=0.4658 auc_roc=0.6894
depth=20 f1=0.4658 auc_roc=0.6894
depth=21 f1=0.4658 auc_roc=0.6894
depth=22 f1=0.4658 auc_roc=0.6894
depth=23 f1=0.4658 auc_roc=0.6894
depth=24 f1=0.4658 auc_roc=0.6894
depth=25 f1=0.4658 auc_roc=0.6894
depth=26 f1=0.4658 auc_roc=0.6894
depth=27 f1=0.4658 auc_roc=0.6894
depth=28 f1=0.4658 auc_roc=0.6894
depth=29 f1=0.4658 auc_roc=0.6894


Уменьшение выборки плохо сказалось на метрике для дерева. Посмотрим на лес.

In [50]:
best_result=0
best_down_forest_model = None
for i in range(1,45):
    model = RandomForestClassifier(random_state = 12345, n_estimators = i)
    model.fit(features_downsampled, target_downsampled)
    predictions_valid = model.predict(features_valid)
    res = f1_auc(model, target_valid, predictions_valid)
    print('n_estimators={:.0f} f1={:>.4f} auc_roc={:>.4f}'.format(i, *res))
    
    if res[0] > best_result:
        best_result = res[0]
        best_down_forest_model = model

n_estimators=1 f1=0.4407 auc_roc=0.6681
n_estimators=2 f1=0.4619 auc_roc=0.7271
n_estimators=3 f1=0.5040 auc_roc=0.7690
n_estimators=4 f1=0.5375 auc_roc=0.7717
n_estimators=5 f1=0.4952 auc_roc=0.7751
n_estimators=6 f1=0.5289 auc_roc=0.7857
n_estimators=7 f1=0.5153 auc_roc=0.7969
n_estimators=8 f1=0.5443 auc_roc=0.8062
n_estimators=9 f1=0.5301 auc_roc=0.8101
n_estimators=10 f1=0.5582 auc_roc=0.8173
n_estimators=11 f1=0.5425 auc_roc=0.8214
n_estimators=12 f1=0.5561 auc_roc=0.8205
n_estimators=13 f1=0.5425 auc_roc=0.8217
n_estimators=14 f1=0.5576 auc_roc=0.8250
n_estimators=15 f1=0.5484 auc_roc=0.8253
n_estimators=16 f1=0.5544 auc_roc=0.8268
n_estimators=17 f1=0.5470 auc_roc=0.8286
n_estimators=18 f1=0.5636 auc_roc=0.8301
n_estimators=19 f1=0.5463 auc_roc=0.8327
n_estimators=20 f1=0.5631 auc_roc=0.8333
n_estimators=21 f1=0.5479 auc_roc=0.8324
n_estimators=22 f1=0.5603 auc_roc=0.8336
n_estimators=23 f1=0.5500 auc_roc=0.8353
n_estimators=24 f1=0.5527 auc_roc=0.8336
n_estimators=25 f1=0.5427

Аналогичная ситуация и для леса - уменьшение не пошло на пользу. Посмотрим регрессию.

In [51]:
down_logistic_model = LogisticRegression(random_state = 12345, solver='liblinear')
down_logistic_model.fit(features_downsampled, target_downsampled)
predictions_valid = down_logistic_model.predict(features_valid)

print('f1={:>.4f} auc_roc={:>.4f}'.format(*f1_auc(down_logistic_model, target_valid, predictions_valid)))

f1=0.4732 auc_roc=0.7735


И для логистической регрессии уменьшение выборки несколько ухудшило показатель метрики.

### Выводы

Провели балансирование классов учётом весов, увеличением и уменьшением выборки. Лучший показатель дало увеличение выборки. Лучшая модель - модель случайного леса. У этой модели единственной добилисть требуемой метрики.


## Тестирование модели

Посмотрим как поведут себя модели на тестовой выборке. Объявим функцию-которая будет определять метрики по тестовой выборке и выводить информацию

In [52]:
def func_test(model):
    predictions_test = model.predict(features_test)
    res = f1_auc(model, target_test, predictions_test)
    print('f1={:>.4f} auc_roc={:>.4f}'.format(*res))
    return None

Будем выводить для модели результаты в порядке обучения моделей 
1) Без балансировки

2) С учётом весов

3) С увеличением выборки

4) с уменьшением выборки

Посмотрим на дерево

In [53]:
func_test(best_tree_model)

f1=0.5620 auc_roc=0.4926


In [54]:
func_test(best_bal_tree_model)

f1=0.5934 auc_roc=0.4941


In [55]:
func_test(best_up_tree_model)

f1=0.5934 auc_roc=0.4941


In [56]:
func_test(best_down_tree_model)

f1=0.5842 auc_roc=0.4933


Наилучший результат на тестировании показало дерево с балансировкой и с увеличением выборки эта одна и та же модель с глубиной 6.

Посмотрим на лес

In [57]:
func_test(best_forest_model)

f1=0.5629 auc_roc=0.4999


In [58]:
func_test(best_bal_forest_model)

f1=0.5524 auc_roc=0.5011


In [59]:
func_test(best_up_forest_model)

f1=0.5990 auc_roc=0.4948


In [60]:
func_test(best_down_forest_model)

f1=0.5817 auc_roc=0.5027


Лучший результат показазал лес с увеличением выборки - несколько выше требуемого. Посмотрим на регрессию.

In [61]:
func_test(logistic_model)

f1=0.3007 auc_roc=0.4938


In [62]:
func_test(bal_logistic_model)

f1=0.5034 auc_roc=0.4927


In [63]:
func_test(up_logistic_model)

f1=0.5029 auc_roc=0.4926


In [64]:
func_test(down_logistic_model)

f1=0.4988 auc_roc=0.4936


Лучший результат показала регрессия при увеличении выборки, но до требуемой метрики не дотягивает.

## Общие выводы 

В ходе исследования проверили баланс классов и обучили модели дерева решений, случайного леса и логистической регрессии без учёта дисбаланса. Лучший результат показал случайный лес. Однако у всех моделей метрика не достигала желаемой.

Для улучшения баланса классов использовали веса, увеличение и уменьшение выборки. Лучший способ - увеличение выборки. Лучший результат опять показал лес на увеличенной выборке. У этой модели единственной на валидации метрика f1 получилась больше 0,59. 

В ходе тестирования лучший результат показал случайный лес с увеличенной выборки с превышением требуемой метрики.
Также модель дерева показала на тестировании метрику больше 0,59. Вообще стоит заметить, что на тестировании у всех моделей значения получились ольше чем на валидации.

Наилучшая модель - модель случайного леса, обученная на увеличенной выборке.

Что касается AUC_ROC - то эта метрика на тестировании оказалась меньше, чем на валидации и кроме того - практически везде меньше, чем 0,5 (как для случайной модели).  Если сравнивать AUC_ROC, то максимумы у метрик в разных ситуация оказались. Вероятно эти метрики используются для различных задач.